### PR TITLE
Feature: Placeholder Token Option

### PR DESCRIPTION
**Pull request by yongkanm**
### Description

A new type of token option (placeholder token option) is implemented in this pull request. This token option type serves as a placeholder to allow the user to integrate Token ATM with other third-party services. In addition to general configurations for all token options, the following configurations  are available for this type of token option:

1.  Optionally configure a start time (with exceptions for sections)
2.  Optionally configure an end time (with exceptions for sections)
3.  Optionally allow multiple approved requests (-1 for unlimited)
4.  Exclude other token options

 A student's request will be approved only if all following conditions met:

1.  The student does not reach the maximum allowed amount of approved requests for this token option. If the amount of multiple requests is unlimited (-1), this condition will always met.
2.  The student does not have any approved request for token options that this token option is excluded with.
3.  If the token balance change for this token option is negative, the student should have sufficient token (i.e., have a non-negative balance after spending for this token option)
4.  If start time is configured, the request should be made no earlier than the start time.
5.  If end time is configured, the request should be made no later than the end time.

When approved, Token ATM will record the request and made corresponding token balance change. The user could then export requests for this token option as csv.

In addition to the new token option type, this pull request also includes an enhancement and a bugfix:

*   Enhancement: When configuring exceptions for sections on start time, end time, and new due date, the time for the exception is set to the current time upon creation. It could be frustrating for the user to manually adjust the date to 00:00:00 or 23:59:59. Therefore, a change is made to make the default time behavior for exceptions the same as the default time for start time/end time/new due time (i.e., 00:00:00 for start time; 23:59:59 for end time and new due time).
*   Bugfix: Now at the start of the request processing, several asynchronous checks such as the publish state of Token ATM Log and whether there are missing credentials are performed. However, the button to start request processing is not disabled during these checks. The user could accidentally clicks the “Start Request Processing” twice, which then causes two request processing instance run simultaneously. This will breaks student's Token ATM record. A change is made to disable the “Start Request Processing” button during these checks.

### Testing Note

1.  Please test if the new type of token option works as described.
2.  Please test if the enhancement and the bugfix mentioned above works as described.

### CODE REVIEW

**Review by epsilonError — DISMISSED** [review]
Testing is passed, the feature, enhancement, and bugfix all work.

There are some comments on code typos and notes on possible User Interaction trouble, but they are not a reason to block the approval of the current pull request.

Feel free to address anything you'd like before merging.

File: src/app/components/request-process/request-process.component.ts [hunk 1]
```diff
@@ -79,29 +83,23 @@ export class RequestProcessComponent implements CourseConfigurable {
                 await this.onRequestProcessingComplete(false);
             }
         });
+        this.isProcessing = false;
     }
 
     public async onRequestProcessingComplete(recogfiure = true): Promise<void> {
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/components/request-process/request-process.component.ts:89]: Thanks for catching that! Fix will be included in the next commit.

File: src/app/components/request-process/request-process.component.ts [hunk 2]
```diff
@@ -79,29 +83,23 @@ export class RequestProcessComponent implements CourseConfigurable {
...
         this.progress = undefined;
         this.message = undefined;
         this.individualProgress = undefined;
         this.individualMessage = undefined;
+        this.isProcessing = true;
         if (this.course && recogfiure) await this.configureCourse(this.course);
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/components/request-process/request-process.component.ts:95]: Thanks for catching that! Fix will be included in the next commit.

File: src/app/token-options/mixins/multiple-requests-mixin.ts
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/token-options/mixins/multiple-requests-mixin.ts:1]: Thanks for catching that! Fix will be included in the next commit.

File: src/app/request-handlers/guards/multiple-requests-guard.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            if (!request.tokenOption) continue;
+            if (!request.isApproved) continue;
+            if (this.tokenOption.id == request.tokenOption.id) {
+                count++;
+            }
+        }
+        if (this.tokenOption.allowedReqeustCnt != -1 && count >= this.tokenOption.allowedReqeustCnt)
+            onReject(
+                this.tokenOption.allowedReqeustCnt != 1
+                    ? `Your already have ${count} approved requests for this token option and cannot have more.`
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/request-handlers/guards/multiple-requests-guard.ts:23]: Thanks for catching that! Fix will be included in the next commit.

File: src/app/token-option-field-component-factories/placeholder-token-option-field-component-factory.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                    }
+                })
+                .transformDest(
+                    async ([tokenOptionData, startTime, endTime, allowedReqeustCnt, excludeTokenOptionIds]) => {
+                        return {
+                            ...tokenOptionData,
+                            type: 'placeholder-token-option',
+                            startTime: startTime ?? null,
+                            endTime: endTime ?? null,
+                            allowedReqeustCnt: allowedReqeustCnt ?? 1,
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/token-option-field-component-factories/placeholder-token-option-field-component-factory.ts:188]: Thanks for catching that! Fix will be included in the next commit.

**Comment by yongkanm** [comment]
> **Note**: Uncaught error during request processing prevents "Start Request Processing" from leaving the `isProcessing` state.
> 
> Reproduce: Setup Canvas course to make an `Error: Multiple Canvas assignments found` error after clicking "Start Request Processing".

Thanks for pointing this out! I would say this is actually the expected behavior since an uncaught error means that the program could possibly be in an inconsistent state, which means that the user should restart it. I believe the fix for this issue is to handle the error properly, but I think it should belong to another pull request.

**Comment by yongkanm** [comment]
> **Note**: When `00:00:00` is supplied as a time, the table in the Request Submission Survey that student's see has the time labelled as `24:00:00`. This could be confusing for students, having the table also display `00:00:00` would be much preferred.

Agreed. This issue has actually existed for a long time. For some reason I choose to use `kk:mm:ss` to format the date at the very beginning, but I cannot find exactly why I did that. Changing the format string to `HH:mm:ss` will resolve this issue. I will include this change in the next commit.

**Comment by yongkanm** [comment]
> **Note**: Based on the form, users may assume that a single Placeholder Token Option can support infinite, unlimited, or large maximum amount of approved requests. But Canvas can, and currently does, limit the number of submission possible to each Request Submission Survey.
> 
> So it's possible that the provided maximum to approve is unattainable, and the Survey responses certainly cap out before infinity.
> 
> This is a constraint of working with Canvas. But I'm unsure how we can make this noticeable from inside Token ATM.

I haven't test if Canvas limits the amount of submissions per student on an ungraded survey, but the current constraint of maximum 100 requests per token option group is actually manually placed by Token ATM. This constraint is to prevent the encrypted student record from exceeding the maximum length for a comment (65,000 characters according to [the documentation](https://community.canvaslms.com/t5/Canvas-Resource-Documents/Canvas-Character-Limits/ta-p/529365)). Implementing paginated comments for encrypted student record to mitigate this issue, but we still need to test if Canvas itself has a constraint.

**Comment by yongkanm** [comment]
Changes are made to fix typos and change date format string. Please note that since a typo in configuration data is fixed, for courses that have placeholder token option(s) created using the previous version, replacing `allowed_reqeust_cnt` with `allowed_request_cnt` is needed.

**Review by epsilonError — APPROVED** [review]
Feature, enhancement, and bugfix all still function after typos and date format strings are fixed.

**Pull request merged**